### PR TITLE
Improve memory usage for RPA

### PR DESCRIPTION
**Pull request by yangdatou**
- outcore direct RPA
- try to raise something when loading libpbc.get_Gv

### CODE REVIEW

**Comment by sunqm** [comment]
Please put the DF-RPA in a separated implementation. I'm not clear if DF-RPA is a more common choice. If that's true, you can make the DF version as the default one. But the old program without DF should not be removed.

**Comment by yangdatou** [comment]
> Please put the DF-RPA in a separated implementation. I'm not clear if DF-RPA is a more common choice. If that's true, you can make the DF version as the default one. But the old program without DF should not be removed.

Hi Qiming, thank you for the comments. The current implementation in pyscf is in fact DF-RPA. (https://github.com/pyscf/pyscf/blob/master/pyscf/gw/rpa.py) The main difference in my code is the outcore version of AO2MO and forming dielectric matrix in aux basis. If you think it is necessary, I would be glad to add RPA without DF.

**Comment by sunqm** [comment]
I see. You don't have to implement more. I will read the code later.

**Comment by tberkel** [comment]
Just confirming what @yangdatou is saying: DF-RPA is really the "standard" implementation of RPA. Without DF, I don't think there's a way to do RPA with less than N^6 scaling.

File: pyscf/gw/rpa.py [hunk 1]
```diff
@@ -36,85 +37,104 @@
...
 
     # Compute total energy
     e_tot = e_hf + e_corr
-
     logger.debug(rpa, '  RPA total energy = %s', e_tot)
     logger.debug(rpa, '  EXX energy = %s, RPA corr energy = %s', e_hf, e_corr)
 
     return e_tot, e_hf, e_corr
 
-def get_rpa_ecorr(rpa, Lpq, freqs, wts):
```
Comment (sunqm) [inline pyscf/gw/rpa.py:85]: Why these methods are removed? Are they problematic?

File: pyscf/pbc/gto/cell.py [hunk 2]
```diff
@@ -589,13 +589,21 @@ def get_Gv_weights(cell, mesh=None, **kwargs):
...
+
+    try:
+        fn(Gv.ctypes.data_as(ctypes.c_void_p),
+        rx.ctypes.data_as(ctypes.c_void_p),
+        ry.ctypes.data_as(ctypes.c_void_p),
+        rz.ctypes.data_as(ctypes.c_void_p),
+        mesh.ctypes.data_as(ctypes.c_void_p),
+        b.ctypes.data_as(ctypes.c_void_p))
+
+    except Exception:
```
Comment (sunqm) [inline pyscf/pbc/gto/cell.py:604]: Why calling the C function can fail? Can this error be more specific?

**Review by sunqm — APPROVED** [review]

File: pyscf/pbc/gto/cell.py (hunk 2 above)
Comment (yangdatou) [reply pyscf/pbc/gto/cell.py:604]: I updated pyscf to master and didn't compile the new binary file. The code breaks in some weird place. Adding this assert might help people (like me) to locate the problem. It shall also be the common practice in pyscf when call C functions.

Comment (sunqm) [reply pyscf/pbc/gto/cell.py:604]: If the symbol not available from the C extension, the code should be failed at the line `fn = getattr(libpbc, 'get_Gv', None)` . You can change it to `fn libpbc.get_Gv` , which will give an accurate error info. Catching errors for the C caller is misleading.

**Pull request merged**